### PR TITLE
jbranchaud/tt 106 make sure buy more seats route create a

### PR DESCRIPTION
**Pull request by jbranchaud**
### feat(tt): handle another bulk purchase scenario

If a user has already purchased a product and then while not
authenticated they go back through the purchase flow, we should be able
to recognize that they have already purchased and create a bulk coupon
for their new purchase that they can share with their team.

Here is the post-purchase screen that they get now if they are making a subsequent individual purchase of the product.

CleanShot 2023-02-14 at 11 30 01@2x

doritos

### CODE REVIEW

File: packages/database/src/prisma-api.ts [hunk 1]
```diff
@@ -394,6 +394,26 @@ export function getSdk(
...
+      // if an existingPurchase is found, but no existingBulkCoupon is found,
+      // then the logic below will account for that and create a new bulk
+      // coupon for the requested quantity.
+      const existingPurchase = await ctx.prisma.purchase.findFirst({
+        where: {
+          productId,
+          userId,
+          status: 'Valid',
+        },
+      })
```
Comment (jbranchaud) [inline packages/database/src/prisma-api.ts:410]: determine if there is an existing purchase of this product which we factor in below to determine if this new purchase should be considered a _bulk_ purchase

File: packages/database/src/prisma-api.ts [hunk 2]
```diff
@@ -394,6 +394,26 @@ export function getSdk(
...
+
+      // TODO: This doesn't seem to be looking up the bulk coupon based
+      // on the product ID which will become an issue when any particular
+      // app has more than one product.
+      // E.g. should probably account for `restrictedToProductId`
+      //
```
Comment (jbranchaud) [inline packages/database/src/prisma-api.ts:416]: I think we have a gap here that hasn't been an issue for single-product apps, but we need to be a little more specific with the query. I'll deal with this in a separate PR.

File: packages/database/src/prisma-api.ts [hunk 3]
```diff
@@ -458,11 +481,12 @@ export function getSdk(
         },
       })
 
+      const oneWeekInMilliseconds = 1000 * 60 * 60 * 24 * 7
       const purchaseUserTransfer = ctx.prisma.purchaseUserTransfer.create({
         data: {
           sourceUserId: userId,
           purchaseId: purchaseId,
-          expiresAt: new Date(Date.now() + 1000 * 60 * 60 * 24 * 7),
+          expiresAt: new Date(Date.now() + oneWeekInMilliseconds),
```
Comment (jbranchaud) [inline packages/database/src/prisma-api.ts:489]: I stared at this expression long enough that I decided it was a good idea to extract it as a descriptive variable name.

File: packages/database/src/prisma-api.ts (hunk 2 above)
Comment (joelhooks) [reply packages/database/src/prisma-api.ts:416]: 👍

File: packages/database/src/prisma-api.ts (hunk 3 above)
Comment (joelhooks) [reply packages/database/src/prisma-api.ts:489]: 🤡

**Review by joelhooks — APPROVED** [review]

**Pull request merged**